dbms/src/Dictionaries/PolygonDictionary.cpp

```diff
@@ -0,0 +1,747 @@
+    }
+    for (auto & point : points)
+    {
+        switch (input_type)
+        {
+            case InputType::MultiPolygon:
+                addNewPointFromMultiPolygon(point, data, offset);
+                break;
+            case InputType::SimplePolygon:
+                addNewPointFromSimplePolygon(point, data, offset);
+                break;
+        }
+    }
+}
+
+SimplePolygonDictionary::SimplePolygonDictionary(
+    const std::string & database_,
+    const std::string & name_,
+    const DictionaryStructure & dict_struct_,
+    DictionarySourcePtr source_ptr_,
+    const DictionaryLifetime dict_lifetime_,
+    InputType input_type_,
+    PointType point_type_)
+    : IPolygonDictionary(database_, name_, dict_struct_, std::move(source_ptr_), dict_lifetime_, input_type_, point_type_)
+{
+}
+
+std::shared_ptr<const IExternalLoadable> SimplePolygonDictionary::clone() const
+{
+    return std::make_shared<SimplePolygonDictionary>(
+            this->database,
+            this->name,
+            this->dict_struct,
+            this->source_ptr->clone(),
+            this->dict_lifetime,
+            this->input_type,
+            this->point_type);
+}
+
+bool SimplePolygonDictionary::find(const Point &point, size_t & id) const
+{
+    bool found = false;
+    double area = 0;
+    for (size_t i = 0; i < (this->polygons).size(); ++i)
+    {
+        if (bg::covered_by(point, (this->polygons)[i]))
+        {
+            double new_area = bg::area((this->polygons)[i]);
+            if (!found || new_area < area)
+            {
+                found = true;
+                id = i;
+                area = new_area;
+            }
+        }
+    }
+    return found;
+}
+
+void registerDictionaryPolygon(DictionaryFactory & factory)
+{
+    auto create_layout = [=](const std::string &,
+                             const DictionaryStructure & dict_struct,
+                             const Poco::Util::AbstractConfiguration & config,
+                             const std::string & config_prefix,
+                             DictionarySourcePtr source_ptr) -> DictionaryPtr
+    {
+        const String database = config.getString(config_prefix + ".database", "");
+        const String name = config.getString(config_prefix + ".name");
+
+        if (!dict_struct.key)
+            throw Exception{"'key' is required for a dictionary of layout 'polygon'", ErrorCodes::BAD_ARGUMENTS};
+        if (dict_struct.key->size() != 1)
+            throw Exception{"The 'key' should consist of a single attribute for a dictionary of layout 'polygon'",
+                            ErrorCodes::BAD_ARGUMENTS};
+        // TODO: Once arrays are fully supported this should be changed to a more reasonable check.
+        const auto type_name = (*dict_struct.key)[0].type->getName();
+        IPolygonDictionary::InputType input_type;
+        IPolygonDictionary::PointType point_type;
+        if (type_name == "Array(Array(Array(Array(Float64))))")
```

Comment:
It's not good to compare type name by string literal. Because it can be changed.
Better ways are:
a) use `typeid_cast`
or
b) create a set of supported types and compare using `IDataType::equals`

Reply:
Chose the second option, since I think it is better to check this before creating the dictionary and I don't know how to use typeid_cast with DataTypePtr.